pkgs/development/python-modules/scipy/default.nix

```diff
@@ -17,46 +20,125 @@
 , pybind11
 , pooch
 , libxcrypt
+, xsimd
+, blas
+, lapack
+# These 2 are used at scipy/meson.build - if you want to override the blas
+# implementation, you'll likely only need to override blas and lapack deps, but
+# this might be needed in some cases.
+, blasName ? blas.pname
+, lapackName ? lapack.pname
 }:
 
-buildPythonPackage rec {
+let
+  datasets = [
+    "ascent"
+    "ecg"
+    "face"
+  ];
+  curl = "${lib.getBin buildPackages.curl}/bin/curl";
+  # Make sure numpy uses the same blas package
+  numpy' = numpy.override {
+    inherit blas;
+  };
+  # Additional cross compilation related properties that scipy reads in scipy/meson.build
+  crossFileScipy = writeText "cross-file-scipy.conf" ''
+    [properties]
+    numpy-include-dir = '${numpy}/${python.sitePackages}/numpy/core/include'
+    pybind11-include-dir = '${pybind11}/${python.sitePackages}/pybind11/include'
+    pythran-include-dir = '${pythran}/${python.sitePackages}/pythran'
+    host-python-path = '${python.interpreter}'
+    host-python-version = '${python.pythonVersion}'
+  '';
+in buildPythonPackage rec {
   pname = "scipy";
-  version = "1.10.1";
+  version = "1.11.1";
   format = "pyproject";
 
-  src = fetchPypi {
-    inherit pname version;
-    hash = "sha256-LPnfuAp7RYm6TEDOdYiYbW1c68VFfK0sKID2vC1C86U=";
+  src = fetchFromGitHub {
+    owner = "scipy";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-Myu1++M9dunipKjfhwBruNgP9/XuwX4HaTmQ9vg44Yc=";
+    # Fetch datasets that are needed for tests
+    postFetch = ''
+      # This directory must stay with this name, as it fits the environment
+      # variable XDG_CACHE_DIR set in `preConfigure`. See more details at:
+      # https://github.com/scipy/scipy/pull/18518#issuecomment-1562350648 And
+      # at: https://github.com/scipy/scipy/pull/17965#issuecomment-1560759962
+      mkdir $out/scipy-data
+    '' + (lib.concatStringsSep "\n" (builtins.map (d:
+      # From some reason, if ssl certificates are not available in this
+      # environment, see also:
+      # https://discourse.nixos.org/t/fetchfromgithub-with-a-postfetch-acts-weird/28444/7
+      "${curl} --insecure https://raw.githubusercontent.com/scipy/dataset-${d}/main/${d}.dat -o $out/scipy-data/${d}.dat"
+    ) datasets));
```

Comment:
You are not supposed to download files here. This feels like a hacky way to do IFD without importing nix code. Can't we use one fetcher per file? How many files do we download?

Reply:
> You are not supposed to download files here.

Why not?

> Can't we use one fetcher per file?

The idea here is that this hook doesn't introduce more hashes to maintain. This also means that `nix-update` easily manages to update this `src` without any `passthru.updateScript`.

> How many files do we download?

3. I'd really like to not introduce 3 more hashes :).

Reply:
`postFetch` should be avoided as it makes it difficult to prefetch. In this case a separate fetch may need to be done instead.

Reply:
> `postFetch` should be avoided as it makes it difficult to prefetch.

What do you mean by "prefetch"? Why not simply set an empty hash and see what is the correct hash in a `nix build`?

Reply:
> What do you mean by "prefetch"?

You can't easily manually reproduce the downloaded files and add them to the nix store. Also if any of those downloads fails, all of them need to be retried. Also if the source is updated and the other files stay the same we wouldn't need to redownload everything. It is probably a better idea to do one hash for each download.

Reply:
@SuperSandro2000 I switched to using 1 hash per file, and I added a `passthru.updateScript` that should aid with updating these hashes. Hence I'm marking this conversation as resolved.